File: Controllers/AdminController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Sofranet.Data;

namespace Sofranet.Controllers;

[Authorize(Roles = "Admin")]
public class AdminController : Controller
{
    private readonly ApplicationDbContext _db;

    public AdminController(ApplicationDbContext db)
    {
        _db = db;
    }

    public async Task<IActionResult> Index()
    {
        var userRole = await _db.Roles.FirstOrDefaultAsync(r => r.Name == "User");
        var catererRole = await _db.Roles.FirstOrDefaultAsync(r => r.Name == "Caterer");

        var totalUsers = userRole == null
            ? 0
            : await _db.UserRoles.CountAsync(ur => ur.RoleId == userRole.Id);
        var totalCaterers = catererRole == null
            ? 0
            : await _db.UserRoles.CountAsync(ur => ur.RoleId == catererRole.Id);

        var totalOrders = await _db.Orders.CountAsync();
        var today = DateTime.UtcNow.Date;
        var todayOrders = await _db.Orders.CountAsync(o => o.CreatedAt >= today);
        var volumeList = await _db.Orders
            .Where(o => o.Status != "cancelled")
            .Select(o => o.TotalAmount)
            .ToListAsync();
        var totalVolume = volumeList.Sum();
        var activeMenu = await _db.MenuItems.CountAsync(m => m.IsAvailable);
        var totalRatings = await _db.Ratings.CountAsync();

        // SQLite decimal Sum desteklemiyor, GroupBy sonrası amounts'u listeleyip in-memory aggregate
        var orderAmounts = await _db.Orders
            .Select(o => new { o.CatererId, o.TotalAmount })
            .ToListAsync();

        var topCaterersRaw = orderAmounts
            .GroupBy(x => x.CatererId)
            .Select(g => new
            {
                CatererId = g.Key,
                OrderCount = g.Count(),
                Revenue = g.Sum(x => x.TotalAmount)
            })
            .OrderByDescending(x => x.OrderCount)
            .Take(5)
            .ToList();

        var catererIds = topCaterersRaw.Select(x => x.CatererId).ToList();
        var catererNames = await _db.Users
            .Where(u => catererIds.Contains(u.Id))
            .ToDictionaryAsync(u => u.Id, u => u.FullName);

        var topCaterers = topCaterersRaw.Select(x => new TopCatererRow
        {
            CatererId = x.CatererId,
            Name = catererNames.TryGetValue(x.CatererId, out var n) ? n : "-",
            OrderCount = x.OrderCount,
            Revenue = x.Revenue
        }).ToList();

        var recentOrders = await _db.Orders
            .Include(o => o.User)
            .Include(o => o.Caterer)
            .OrderByDescending(o => o.CreatedAt)
            .Take(10)
            .Select(o => new AdminRecentOrder
            {
                Id = o.Id,
                TotalAmount = o.TotalAmount,
                Status = o.Status,
                CreatedAt = o.CreatedAt,
                UserName = o.User!.FullName,
                CatererName = o.Caterer!.FullName
            })
            .ToListAsync();

        ViewBag.TotalUsers = totalUsers;
        ViewBag.TotalCaterers = totalCaterers;
        ViewBag.TotalOrders = totalOrders;
        ViewBag.TodayOrders = todayOrders;
        ViewBag.TotalVolume = totalVolume;
        ViewBag.ActiveMenu = activeMenu;
        ViewBag.TotalRatings = totalRatings;
        ViewBag.TopCaterers = topCaterers;
        ViewBag.RecentOrders = recentOrders;

        ViewData["Title"] = "Yönetici Paneli";
        return View();
    }

    // === USERS ===
    public async Task<IActionResult> Users(string? q, string? role, int page = 1)
    {
        if (page < 1) page = 1;
        var query = _db.Users.AsQueryable();

        if (!string.IsNullOrWhiteSpace(q))
        {
            var ql = q.ToLower();
            query = query.Where(u => u.FullName.ToLower().Contains(ql) || u.Email!.ToLower().Contains(ql));
        }

        if (!string.IsNullOrWhiteSpace(role))
        {
            var roleEntity = await _db.Roles.FirstOrDefaultAsync(r => r.Name == role);
            if (roleEntity != null)
            {
                var idsInRole = await _db.UserRoles
                    .Where(ur => ur.RoleId == roleEntity.Id)
                    .Select(ur => ur.UserId)
                    .ToListAsync();
                query = query.Where(u => idsInRole.Contains(u.Id));
            }
        }

        var total = await query.CountAsync();
        var users = await query
            .OrderByDescending(u => u.CreatedAt)
            .Skip((page - 1) * 20).Take(20)
            .Select(u => new AdminUserRow
            {
                Id = u.Id,
                FullName = u.FullName,
                Email = u.Email ?? "-",
                PhoneNumber = u.PhoneNumber,
                Address = u.Address,
                CreatedAt = u.CreatedAt
            })
            .ToListAsync();

        var userIds = users.Select(u => u.Id).ToList();
        var rolesByUser = await (from ur in _db.UserRoles
                                 join r in _db.Roles on ur.RoleId equals r.Id
                                 where userIds.Contains(ur.UserId)
                                 select new { ur.UserId, RoleName = r.Name })
                                 .ToListAsync();
        var roleMap = rolesByUser.GroupBy(x => x.UserId)
            .ToDictionary(g => g.Key, g => g.Select(x => x.RoleName).FirstOrDefault() ?? "-");

        foreach (var u in users)
        {
            u.Role = roleMap.TryGetValue(u.Id, out var r) ? (r ?? "-") : "-";
        }

        ViewBag.Users = users;
        ViewBag.Query = q;
        ViewBag.RoleFilter = role;
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / 20.0);
        ViewBag.Total = total;
        ViewData["Title"] = "Kullanıcılar";
        return View();
    }

    // === CATERERS ===
    public async Task<IActionResult> Caterers(string? q, int page = 1)
    {
        if (page < 1) page = 1;
        var catererRole = await _db.Roles.FirstOrDefaultAsync(r => r.Name == "Caterer");
        var display = new List<AdminCatererRow>();
        int total = 0;

        if (catererRole != null)
        {
            var catererIds = await _db.UserRoles
                .Where(ur => ur.RoleId == catererRole.Id)
                .Select(ur => ur.UserId).ToListAsync();

            var query = _db.Users.Where(u => catererIds.Contains(u.Id));
            if (!string.IsNullOrWhiteSpace(q))
            {
                var ql = q.ToLower();
                query = query.Where(u => u.FullName.ToLower().Contains(ql) || u.Email!.ToLower().Contains(ql));
            }

            total = await query.CountAsync();
            var caterers = await query
                .OrderBy(u => u.FullName)
                .Skip((page - 1) * 20).Take(20)
                .ToListAsync();

            foreach (var c in caterers)
            {
                var orderCount = await _db.Orders.CountAsync(o => o.CatererId == c.Id);
                var amounts = await _db.Orders
                    .Where(o => o.CatererId == c.Id && o.Status != "cancelled")
                    .Select(o => o.TotalAmount).ToListAsync();
                var revenue = amounts.Sum();

                var ratings = await _db.Ratings
                    .Where(r => r.CatererId == c.Id && r.Order!.Status == "completed")
                    .Select(r => (decimal)r.CatererRating).ToListAsync();
                decimal? avg = ratings.Count > 0 ? Math.Round(ratings.Average(), 2) : null;

                display.Add(new AdminCatererRow
                {
                    Id = c.Id,
                    FullName = c.FullName,
                    Email = c.Email ?? "-",
                    PhoneNumber = c.PhoneNumber,
                    Address = c.Address,
                    OrderCount = orderCount,
                    Revenue = revenue,
                    AvgRating = avg,
                    RatingCount = ratings.Count
                });
            }
        }

        ViewBag.Caterers = display;
        ViewBag.Query = q;
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / 20.0);
        ViewBag.Total = total;
        ViewData["Title"] = "Restoranlar";
        return View();
    }

    // === ORDERS ===
    public async Task<IActionResult> Orders(string? status, DateTime? dateFrom, DateTime? dateTo,
        string? q, int page = 1)
    {
        if (page < 1) page = 1;
        var query = _db.Orders
            .Include(o => o.User)
            .Include(o => o.Caterer)
            .AsQueryable();

        if (!string.IsNullOrEmpty(status))
            query = query.Where(o => o.Status == status);

        if (dateFrom.HasValue)
            query = query.Where(o => o.CreatedAt >= dateFrom.Value);

        if (dateTo.HasValue)
        {
            var to = dateTo.Value.Date.AddDays(1);
            query = query.Where(o => o.CreatedAt < to);
        }

        if (!string.IsNullOrWhiteSpace(q))
        {
            var ql = q.ToLower();
            query = query.Where(o =>
                o.User!.FullName.ToLower().Contains(ql) ||
                o.User!.Email!.ToLower().Contains(ql) ||
                o.Caterer!.FullName.ToLower().Contains(ql) ||
                o.Caterer!.Email!.ToLower().Contains(ql));
        }

        var total = await query.CountAsync();
        var orders = await query
            .OrderByDescending(o => o.CreatedAt)
            .Skip((page - 1) * 20).Take(20)
            .Select(o => new AdminOrderRow
            {
                Id = o.Id,
                TotalAmount = o.TotalAmount,
                Status = o.Status,
                CreatedAt = o.CreatedAt,
                UserName = o.User!.FullName,
                UserEmail = o.User!.Email ?? "-",
                CatererName = o.Caterer!.FullName
            })
            .ToListAsync();

        ViewBag.Orders = orders;
        ViewBag.Status = status;
        ViewBag.DateFrom = dateFrom?.ToString("yyyy-MM-dd");
        ViewBag.DateTo = dateTo?.ToString("yyyy-MM-dd");
        ViewBag.Query = q;
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / 20.0);
        ViewBag.Total = total;
        ViewData["Title"] = "Tüm Siparişler";
        return View();
    }

    // === LOGS ===
    // "action" parametre ismi MVC route'unda reserved (action method adı route value'sundan alınır).
    // Bu nedenle "actionType" kullanıyoruz, view'da da name="actionType".
    public async Task<IActionResult> Logs(string? actionType, string? q, string? ip,
        DateTime? dateFrom, DateTime? dateTo, int page = 1)
    {
        if (page < 1) page = 1;
        var query = _db.LogEntries.AsQueryable();

        if (!string.IsNullOrEmpty(actionType))
            query = query.Where(l => l.Action == actionType);

        if (!string.IsNullOrEmpty(ip))
            query = query.Where(l => l.IpAddress != null && l.IpAddress.Contains(ip));

        if (dateFrom.HasValue)
            query = query.Where(l => l.CreatedAt >= dateFrom.Value);

        if (dateTo.HasValue)
        {
            var to = dateTo.Value.Date.AddDays(1);
            query = query.Where(l => l.CreatedAt < to);
        }

        if (!string.IsNullOrWhiteSpace(q))
        {
            var ql = q.ToLower();
            var matchingUserIds = await _db.Users
                .Where(u => u.FullName.ToLower().Contains(ql) || u.Email!.ToLower().Contains(ql))
                .Select(u => u.Id).ToListAsync();
            query = query.Where(l => l.UserId != null && matchingUserIds.Contains(l.UserId));
        }

        var total = await query.CountAsync();
        var rawLogs = await query
            .OrderByDescending(l => l.CreatedAt)
            .Skip((page - 1) * 50).Take(50)
            .ToListAsync();
        var logs = rawLogs.Select(l => new AdminLogRow
        {
            Id = l.Id,
            Action = l.Action,
            Details = l.Details ?? string.Empty,
            IpAddress = l.IpAddress ?? string.Empty,
            CreatedAt = l.CreatedAt,
            UserId = l.UserId
        }).ToList();

        var logUserIds = logs.Where(l => !string.IsNullOrEmpty(l.UserId))
            .Select(l => l.UserId!).Distinct().ToList();
        var userInfoMap = await _db.Users
            .Where(u => logUserIds.Contains(u.Id))
            .Select(u => new { u.Id, u.Email, u.FullName })
            .ToListAsync();
        var userDict = userInfoMap.ToDictionary(u => u.Id, u => new { u.Email, u.FullName });

        foreach (var l in logs)
        {
            if (l.UserId != null && userDict.TryGetValue(l.UserId, out var info))
            {
                l.UserEmail = info.Email;
                l.UserName = info.FullName;
            }
        }

        var actions = await _db.LogEntries
            .Select(l => l.Action)
            .Distinct()
            .OrderBy(a => a)
            .ToListAsync();

        ViewBag.Logs = logs;
        ViewBag.Actions = actions;
        ViewBag.ActionFilter = actionType;
        ViewBag.Query = q;
        ViewBag.IpFilter = ip;
        ViewBag.DateFrom = dateFrom?.ToString("yyyy-MM-dd");
        ViewBag.DateTo = dateTo?.ToString("yyyy-MM-dd");
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / 50.0);
        ViewBag.Total = total;
        ViewData["Title"] = "Sistem Logları";
        return View();
    }
}

public class AdminUserRow
{
    public string Id { get; set; } = "";
    public string FullName { get; set; } = "";
    public string Email { get; set; } = "";
    public string? PhoneNumber { get; set; }
    public string? Address { get; set; }
    public DateTime CreatedAt { get; set; }
    public string Role { get; set; } = "-";
}

public class AdminCatererRow
{
    public string Id { get; set; } = "";
    public string FullName { get; set; } = "";
    public string Email { get; set; } = "";
    public string? PhoneNumber { get; set; }
    public string? Address { get; set; }
    public int OrderCount { get; set; }
    public decimal Revenue { get; set; }
    public decimal? AvgRating { get; set; }
    public int RatingCount { get; set; }
}

public class AdminOrderRow
{
    public int Id { get; set; }
    public decimal TotalAmount { get; set; }
    public string Status { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public string UserName { get; set; } = "";
    public string UserEmail { get; set; } = "";
    public string CatererName { get; set; } = "";
}

public class AdminLogRow
{
    public long Id { get; set; }
    public string Action { get; set; } = "";
    public string Details { get; set; } = "";
    public string IpAddress { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public string? UserId { get; set; }
    public string? UserEmail { get; set; }
    public string? UserName { get; set; }
}

public class TopCatererRow
{
    public string CatererId { get; set; } = "";
    public string Name { get; set; } = "-";
    public int OrderCount { get; set; }
    public decimal Revenue { get; set; }
}

public class AdminRecentOrder
{
    public int Id { get; set; }
    public decimal TotalAmount { get; set; }
    public string Status { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public string UserName { get; set; } = "";
    public string CatererName { get; set; } = "";
}

File: controllers/adminController.js
// admin controller - dashboard + yonetim sayfalari
const db = require('../config/db');
const { getPagination } = require('../utils/pagination');

async function dashboard(req, res) {
  try {
    const statsRes = await db.query(`
      SELECT
        (SELECT COUNT(*) FROM users WHERE role = 'user') as total_users,
        (SELECT COUNT(*) FROM users WHERE role = 'caterer') as total_caterers,
        (SELECT COUNT(*) FROM orders) as total_orders,
        (SELECT COUNT(*) FROM orders WHERE created_at::date = CURRENT_DATE) as today_orders,
        (SELECT COALESCE(SUM(total_amount), 0) FROM orders WHERE status != 'cancelled') as total_volume,
        (SELECT COUNT(*) FROM menu_items WHERE is_available = TRUE) as active_menu_items,
        (SELECT COUNT(*) FROM ratings) as total_ratings
    `);

    const topRes = await db.query(`
      SELECT u.id, u.name, u.email,
        COUNT(o.id) as order_count,
        COALESCE(SUM(o.total_amount), 0) as revenue
      FROM users u
      LEFT JOIN orders o ON o.caterer_id = u.id
      WHERE u.role = 'caterer'
      GROUP BY u.id, u.name, u.email
      ORDER BY order_count DESC LIMIT 5
    `);

    const recentRes = await db.query(`
      SELECT o.id, o.total_amount, o.status, o.created_at,
        cu.name as user_name, cu.email as user_email,
        cc.name as caterer_name
      FROM orders o
      JOIN users cu ON cu.id = o.user_id
      JOIN users cc ON cc.id = o.caterer_id
      ORDER BY o.created_at DESC LIMIT 10
    `);

    res.render('admin/dashboard', {
      title: 'Admin Dashboard',
      stats: statsRes.rows[0],
      topCaterers: topRes.rows,
      recentOrders: recentRes.rows
    });
  } catch (err) {
    console.error('admin dashboard error:', err);
    res.status(500).send('Dashboard yüklenemedi');
  }
}

async function usersList(req, res) {
  try {
    const filters = {
      q: req.query.q || '',
      role: req.query.role || ''
    };
    const page = parseInt(req.query.page, 10) || 1;

    const where = [];
    const params = [];
    if (filters.q) {
      params.push('%' + filters.q + '%');
      where.push('(name ILIKE $' + params.length + ' OR email ILIKE $' + params.length + ')');
    }
    if (['user', 'caterer', 'admin'].includes(filters.role)) {
      params.push(filters.role);
      where.push('role = $' + params.length);
    }
    const whereSql = where.length > 0 ? 'WHERE ' + where.join(' AND ') : '';

    const countRes = await db.query('SELECT COUNT(*) FROM users ' + whereSql, params);
    const pag = getPagination(page, 20, parseInt(countRes.rows[0].count, 10));

    params.push(pag.limit);
    params.push(pag.offset);
    const listRes = await db.query(
      `SELECT id, name, email, role, phone, address, created_at
       FROM users ${whereSql}
       ORDER BY created_at DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );

    res.render('admin/users', {
      title: 'Kullanıcılar',
      users: listRes.rows,
      filters,
      pagination: pag,
      currentQuery: req.query
    });
  } catch (err) {
    console.error('admin users list error:', err);
    res.status(500).send('Liste yüklenemedi');
  }
}

async function caterersList(req, res) {
  try {
    const filters = { q: req.query.q || '' };
    const page = parseInt(req.query.page, 10) || 1;

    const where = ["u.role = 'caterer'"];
    const params = [];
    if (filters.q) {
      params.push('%' + filters.q + '%');
      where.push('(u.name ILIKE $' + params.length + ' OR u.email ILIKE $' + params.length + ')');
    }
    const whereSql = 'WHERE ' + where.join(' AND ');

    const countRes = await db.query('SELECT COUNT(*) FROM users u ' + whereSql, params);
    const pag = getPagination(page, 20, parseInt(countRes.rows[0].count, 10));

    params.push(pag.limit);
    params.push(pag.offset);
    const listRes = await db.query(
      `SELECT u.id, u.name, u.email, u.phone, u.address,
        (SELECT COUNT(*) FROM orders WHERE caterer_id = u.id) as order_count,
        (SELECT COALESCE(SUM(total_amount), 0) FROM orders WHERE caterer_id = u.id AND status = 'completed') as revenue,
        (SELECT AVG(caterer_rating)::numeric(3,2) FROM ratings WHERE caterer_id = u.id) as avg_rating
       FROM users u
       ${whereSql}
       ORDER BY u.created_at DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );

    res.render('admin/caterers', {
      title: 'Caterer\'lar',
      caterers: listRes.rows,
      filters,
      pagination: pag,
      currentQuery: req.query
    });
  } catch (err) {
    console.error('admin caterers list error:', err);
    res.status(500).send('Liste yüklenemedi');
  }
}

async function ordersList(req, res) {
  try {
    const filters = {
      status: req.query.status || '',
      date_from: req.query.date_from || '',
      date_to: req.query.date_to || '',
      q: req.query.q || ''
    };
    const page = parseInt(req.query.page, 10) || 1;

    const where = [];
    const params = [];

    if (['pending', 'preparing', 'completed', 'cancelled'].includes(filters.status)) {
      params.push(filters.status);
      where.push('o.status = $' + params.length);
    }
    if (filters.date_from) {
      params.push(filters.date_from);
      where.push('o.created_at >= $' + params.length);
    }
    if (filters.date_to) {
      params.push(filters.date_to);
      where.push('o.created_at <= ($' + params.length + ')::date + INTERVAL \'1 day\'');
    }
    if (filters.q) {
      params.push('%' + filters.q + '%');
      const p = params.length;
      where.push('(cu.name ILIKE $' + p + ' OR cu.email ILIKE $' + p + ' OR cc.name ILIKE $' + p + ' OR cc.email ILIKE $' + p + ')');
    }
    const whereSql = where.length > 0 ? 'WHERE ' + where.join(' AND ') : '';

    const countRes = await db.query(
      `SELECT COUNT(*) FROM orders o
       JOIN users cu ON cu.id = o.user_id
       JOIN users cc ON cc.id = o.caterer_id
       ${whereSql}`,
      params
    );
    const pag = getPagination(page, 20, parseInt(countRes.rows[0].count, 10));

    params.push(pag.limit);
    params.push(pag.offset);
    const listRes = await db.query(
      `SELECT o.id, o.total_amount, o.status, o.created_at,
        cu.name as user_name, cu.email as user_email,
        cc.name as caterer_name
       FROM orders o
       JOIN users cu ON cu.id = o.user_id
       JOIN users cc ON cc.id = o.caterer_id
       ${whereSql}
       ORDER BY o.created_at DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );

    res.render('admin/orders', {
      title: 'Tüm Siparişler',
      orders: listRes.rows,
      filters,
      pagination: pag,
      currentQuery: req.query
    });
  } catch (err) {
    console.error('admin orders list error:', err);
    res.status(500).send('Liste yüklenemedi');
  }
}

async function logsList(req, res) {
  try {
    const filters = {
      action: req.query.action || '',
      q: req.query.q || '',
      ip: req.query.ip || '',
      date_from: req.query.date_from || '',
      date_to: req.query.date_to || ''
    };
    const page = parseInt(req.query.page, 10) || 1;

    const where = [];
    const params = [];

    if (filters.action) {
      params.push(filters.action);
      where.push('l.action = $' + params.length);
    }
    if (filters.q) {
      params.push('%' + filters.q + '%');
      where.push('(u.email ILIKE $' + params.length + ' OR u.name ILIKE $' + params.length + ')');
    }
    if (filters.ip) {
      params.push('%' + filters.ip + '%');
      where.push('l.ip_address LIKE $' + params.length);
    }
    if (filters.date_from) {
      params.push(filters.date_from);
      where.push('l.created_at >= $' + params.length);
    }
    if (filters.date_to) {
      params.push(filters.date_to);
      where.push('l.created_at <= ($' + params.length + ')::date + INTERVAL \'1 day\'');
    }
    const whereSql = where.length > 0 ? 'WHERE ' + where.join(' AND ') : '';

    const countRes = await db.query(
      `SELECT COUNT(*) FROM logs l LEFT JOIN users u ON u.id = l.user_id ${whereSql}`,
      params
    );
    const pag = getPagination(page, 50, parseInt(countRes.rows[0].count, 10));

    params.push(pag.limit);
    params.push(pag.offset);
    const listRes = await db.query(
      `SELECT l.id, l.action, l.details, l.ip_address, l.created_at,
        u.email as user_email, u.name as user_name, u.role as user_role
       FROM logs l
       LEFT JOIN users u ON u.id = l.user_id
       ${whereSql}
       ORDER BY l.created_at DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );

    // dropdown icin distinct action listesi
    const actionsRes = await db.query(
      'SELECT DISTINCT action FROM logs ORDER BY action'
    );

    res.render('admin/logs', {
      title: 'Sistem Logları',
      logs: listRes.rows,
      actions: actionsRes.rows.map(r => r.action),
      filters,
      pagination: pag,
      currentQuery: req.query
    });
  } catch (err) {
    console.error('admin logs list error:', err);
    res.status(500).send('Loglar yüklenemedi');
  }
}

module.exports = { dashboard, usersList, caterersList, ordersList, logsList };
